### ISSUE TITLE
[wasm-mt] Publish `dotnet.worker.js`

### ISSUE THREAD

**Issue by lambdageek** [dotnet/runtime#68509]
The `WasmApp.targets`  should place `dotnet.worker.js` next to `dotnet.js` 

There may also be an issue with Blazor `dotnet publish` doing likewise.

**Comment by pavelsavara** [comment]
We probably need a way how to customize it. 
- so that it's compatible with ES6 when loading the `dotnet.js`
- so that it's compatible our module exports, which are not `Module` object
- so that it's CSP compliant, no `eval`
- so that it doesn't use `require` in NodeJs + ES6

**Comment by lewing** [comment]
I'd love to have consistent naming pattern for all the workers and we need to make sure that the blazor publish step includes them in the output.

cc @radical @javiercn

**Comment by radical** [comment]
@javiercn if we populate a msbuild item with the files to deploy (not exhaustive), then would you be able to use that to ensure that they get deployed wherever needed?

**Comment by javiercn** [comment]
@radical yes. that would help.

@lewing I set up some time with Ankit to go through the details.

**Comment by lewing** [comment]
It is worth reading through https://github.com/dotnet/runtime/issues/48016 as well

**Comment by lambdageek** [comment]
> I'd love to have consistent naming pattern for all the workers and we need to make sure that the blazor publish step includes them in the output.

renaming `dotnet.worker.js` is possible, if that's what we want to do - it looks like Emscripten will go through `locateFile` to get the URL to load.  (going the other way - when `dotnet.worker.js` imports `dotnet.js` - it goes through the `mainUrlOrBlob` property on the module - which we do set up : `anyModule.mainScriptUrlOrBlob = replacements.scriptUrl; // this is needed by worker threads`

tl;dr if we want the published app to use different/consistent naming for worker JS files, we can do that.

**Comment by lambdageek** [comment]
We're going to need to install a replacement for `PThread.allocateUnusedWorker` (here's the default one):

```javascript
 allocateUnusedWorker: function() {
  if (!Module["locateFile"]) {
   PThread.unusedWorkers.push(new Worker(new URL("dotnet.worker.js", import.meta.url)));
   return;
  }
  var pthreadMainJs = locateFile("dotnet.worker.js");
  PThread.unusedWorkers.push(new Worker(pthreadMainJs));
 },
 getNewWorker: function() {
  if (PThread.unusedWorkers.length == 0) {
   return;
   PThread.allocateUnusedWorker();
   PThread.loadWasmModuleToWorker(PThread.unusedWorkers[0]);
  }
  return PThread.unusedWorkers.pop();
 }
};
```

We should use the new `js-module-threads` asset type and call`resolve_module_url` to get the worker URL. see https://github.com/dotnet/runtime/pull/73073#issuecomment-1201434738

**Comment by lambdageek** [comment]
Replacing `Module["PThread"]["init"]` is not really feasible - emscripten calls `PThread.init` very early - before our replacement code runs.

On the bright side `PThread.init` just calls `PThread.initMainThread` or `PThread.initWorker`.  `initWorker` is fine as is.  `initMainThread` doesn't do anything other than call `allocateUnusedWorker` based on the size of `PTHREAD_POOL_SIZE`.

So we could:
1. set `PTHREAD_POOL_SIZE` to `0`
2. make our own `pthreadPreallocatePool` function that we can call during startup, based on some `MONO_PTHREAD_POOL_SIZE` property that we define.
3. replace `allocateUnusedWorker`  with one that uses `resolve_asset_path` (from `assets.ts`).  the replacement is too late for `PThread.init`, but that doesn't matter since it won't be allocating any workers.

**Comment by pavelsavara** [comment]
The way how to resolve the asset is now merged
```
        const asset = resolve_asset_path("js-module-threads");
        mono_assert(asset && asset.resolvedUrl, "Can't find js-module-threads");
        const worker = new Worker(asset.resolvedUrl);
```

You will also need to add it to the mono-config.json via `src\tasks\WasmAppBuilder\WasmAppBuilder.cs`
```
config.Assets.Add(new ThreadsWorkerEntry ("dotnet.worker.js") );
```

**Comment by pavelsavara** [comment]
This is what SDK needs to do https://github.com/dotnet/sdk/pull/26966/files

**Comment by lambdageek** [comment]
yep, I have it basically working. PR soon
